src/PlusDataCollection/Atracsys/vtkPlusAtracsysTracker.cxx

```diff
@@ -0,0 +1,396 @@
+  if (this->Internal->ActiveMarkerPairingTimeSec > 60)
+  {
+    LOG_WARNING("Are you sure you want to leave active marker pairing enabled for more than 60 seconds?");
+  }
+
+  enum TRACKING_TYPE
+  {
+    ACTIVE,
+    PASSIVE
+  };
+
+  XML_FIND_NESTED_ELEMENT_REQUIRED(dataSourcesElement, deviceConfig, "DataSources");
+  for (int nestedElementIndex = 0; nestedElementIndex < dataSourcesElement->GetNumberOfNestedElements(); nestedElementIndex++)
+  {
+    vtkXMLDataElement* toolDataElement = dataSourcesElement->GetNestedElement(nestedElementIndex);
+    if (STRCASECMP(toolDataElement->GetName(), "DataSource") != 0)
+    {
+      // if this is not a data source element, skip it
+      continue;
+    }
+    if (toolDataElement->GetAttribute("Type") != NULL && STRCASECMP(toolDataElement->GetAttribute("Type"), "Tool") != 0)
+    {
+      // if this is not a Tool element, skip it
+      continue;
+    }
+    std::string toolId(toolDataElement->GetAttribute("Id"));
+    if (toolId.empty())
+    {
+    // tool doesn't have ID needed to generate transform
+    LOG_ERROR("Failed to initialize Atracsys tool: DataSource Id is missing.");
+    continue;
+    }
+
+    TRACKING_TYPE toolTrackingType;
+    XML_READ_ENUM2_ATTRIBUTE_NONMEMBER_OPTIONAL(TrackingType, toolTrackingType, toolDataElement, "ACTIVE", ACTIVE, "PASSIVE", PASSIVE);
+    if (toolTrackingType == ACTIVE)
+    {
+      // active tool, can be loaded directly into FtkGeometryIdMappedToToolId
+      int ftkGeometryId = -1;
+      XML_READ_SCALAR_ATTRIBUTE_NONMEMBER_OPTIONAL(int, GeometryId, ftkGeometryId, toolDataElement);
+      if (ftkGeometryId != -1)
+      {
+        std::pair<int, std::string> thisTool(ftkGeometryId, toolId);
+        this->Internal->FtkGeometryIdMappedToToolId.insert(thisTool);
+      }
+      else
+      {
+        LOG_ERROR("Active tool with Id " << toolId << " is missing a GeometryId");
+        return PLUS_FAIL;
+      }
+    }
+    else if (toolTrackingType == PASSIVE)
+    {
+      // passive tool, load into PlusIdMappedToGeometryFilename to have geometry loaded in InternalConnect
+      const char* geometryFile;
+      if ((geometryFile = toolDataElement->GetAttribute("GeometryFile")) != NULL)
+      {
+        std::pair<std::string, std::string> thisTool(toolId, geometryFile);
+        this->Internal->PlusIdMappedToGeometryFilename.insert(thisTool);
+      }
+      else
+      {
+        LOG_ERROR("Passive tool with Id " << toolId << " is missing GeometryFile (.ini).");
+        return PLUS_FAIL;
+      }
+    }
+  }
+
+  return PLUS_SUCCESS;
+}
+
+//----------------------------------------------------------------------------
+PlusStatus vtkPlusAtracsysTracker::WriteConfiguration(vtkXMLDataElement* rootConfigElement)
+{
+  LOG_TRACE("vtkPlusAtracsysTracker::WriteConfiguration");
+  XML_FIND_DEVICE_ELEMENT_REQUIRED_FOR_WRITING(deviceConfig, rootConfigElement);
+  return PLUS_SUCCESS;
```

Comment:
Write out all elements that may be changed using Set...() methods

Reply:
Not sure what you mean here - could you point me to an example please?

Reply:
In general, attributes that you read from XML should be written, too, especially if there is public API to change them.